core/directory.go

```diff
@@ -258,14 +258,66 @@ func (dir *Directory) Entries(ctx context.Context, bk *buildkit.Client, svcs *Se
 
 	entries, err := ref.ReadDir(ctx, bkgw.ReadDirRequest{
 		Path: src,
+		// We do not use IncludePattern here because it also ignores directories.
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	// Filter entries if include is set, we do not ignore directories in order to support recursive option.
+	// It replaces IncludePattern.
```

Comment:
I don't understand why we don't just use `IncludePatterns` here?

It doesn't recurse, so we'd still need multiple calls to it - but the directories should still be returned.

Reply:
We'd also need to update tests: in both tests `*.md` should be `**/*.md` to represent all markdown files in all subdirectories. `*.md` should only match top-level markdown files (in keeping with shell tradition).

Reply:
IncludePatterns only returns directories that contain files that are matched by the glob. e.g. for a pattern `**/*.md` a directory `dir` will only be returned if it (or a child of it) contains a `.md` file. Is this what you're referring to?

IMO `*.md` should never match `foo/bar/*.md`. You should need the pattern `**/*.md` to catch it. Essentially, the glob should probably behave exactly as it does in bash/zsh/all common implementations of sh.

Reply:
You're right! Fixed in [4364df0](https://github.com/dagger/dagger/pull/5824/commits/4364df05a7849233a953d8c94155e72a985a08cc) to use `IncludePattern`